Replace `-e CMD` by positional arguments

coldfix opened the pull request:
This implements perfect forwarding of command line arguments as requested in #308 with the solution suggested by you in #310.

Note that it *does* use `G_OPTION_REMAINING` rather than relying on the remaining `argv`. The reason is simplicity: [`g_option_context_parse`](https://developer.gnome.org/glib/stable/glib-Commandline-option-parser.html#g-option-context-parse) does not consistently strip the `--` from `argv` which would require additional post-processing:

> A '--' option is stripped from argv unless there are unparsed options before and after it, or some of the options after it start with '-'.

The patch is mostly adopted from #310 - with one notable exception: use `G_OPTION_ARG_FILENAME_ARRAY` instead of `G_OPTION_ARG_STRING_ARRAY` to prevent decoding and enable perfect forwarding:

> On UNIX systems, the argv that is passed to main() has no particular encoding, even to the extent that different parts of it may have different encodings. In general, normal arguments and flags will be in the current locale and filenames should be considered to be opaque byte strings. Proper use of `G_OPTION_ARG_FILENAME` vs `G_OPTION_ARG_STRING` is therefore important.

(Feel free to cherry-pick any commit even if not accepting the rest)

Best, Thomas

coldfix committed:
Use strict posix mode to prevent accidental capture of option-arguments

Prevent accidental capture of arguments when `--` is not used, e.g.:

    termite python -c 'help()'

MachFour commented:
I have built this branch and I really like how it works. Thank you Thomas!

dkasak commented:
Yes, please don't remove the `-e` option completely; I'm arriving to this PR from a use case ([rofi-sensible-terminal](https://github.com/DaveDavenport/rofi/blob/799ace6dfd7a9c862e3dab011864658c187caae2/script/rofi-sensible-terminal)) which relies on this convention.

However, `-e` in termite works quite a bit differently than it works in most terminals, for instance xterm. While xterm allows passing many arguments after `-e`, termite expects a single monolithic quoted argument. If I understand this PR correctly, it is trying to support the "many arguments" variant of argument passing, just without a flag. If this is the case, why not simply make `-e` a no-op? This way, it wouldn't hurt if scripts passed it, but it would work just as it works without it.

josswright commented:
Can I just add a vote for this PR? I found this having just switched to termite from urxvt, and this was necessary to get the NVim-R plugin working again. I'd already managed to work around the behaviour in a few other scripts, but this made it work as a drop-in replacement.

frebib commented:
Can I propose syntax something like the following be used to support both single and multiple argument configurations? Ideally for compatibility with most things, it would be good to support both:
`termite -e -- program --arg=1 -a -v`

coldfix commented:
Personally, I'd leave `-e` as is for backward compatibility, but the maintainer requested specifically to drop it.

dkasak commented:
Question for the maintainer: what is the rationale for dropping `-e` completely instead of making it do nothing? It seems to me this will only introduce interoperability pain without any gain.

coldfix committed:
Execute command given as positional arguments

coldfix committed:
Remove '-e' (in favour of positional arguments)

coldfix commented:
Note: I have rebased to resolve merge conflicts and split the removal of `-e` into the final commit - so the first commits can be merged in a backward compatible manner.

josswright commented:
Is there anything preventing this pull request from being accepted?

jelly commented:
@josswright well yes, the time I can spend on termite.

josswright commented:
@jelly Which is very much appreciated!

jelly commented:
Finally looking at this PR, I like the new syntax, but the breakage would be annoying. I'm wondering about a migration path, although dealing with both options would be annoying (i.e.having them both supported at the same time).

@coldfix Sorry for the long delay, but I'd want to merge this somewhere this week, do you think you can rebase it?

coldfix commented:
Hi @jelly , I have rebased. If you want to keep backward compatibility for now, we can leave the `-e` option untouched, i.e. only merge the first two commits.

jelly commented:
Awesome, thanks for the rebase, I'll look at it later this week!

mkaito commented:
I think you need to remove the reference to `execute` [here](https://github.com/coldfix/termite/blob/2b72d9a8844636281e8b90d2b54252c0dd06a4b1/termite.cc#L1762) or it won't compile.

coldfix commented:
> I think you need to remove the reference to execute here or it won't compile.

Thanks! I didn't check after rebasing...

> After testing, I've found that it doesn't work as I expected it to.

Your two snippets do different things.

```
rxvt -e zsh -c weechat-curses
```

Loads termite with a different config file and then runs `zsh`.


```
termite -- zsh -c weechat-curses
```

Runs `zsh -c weechat-curses`, i.e. instructing zsh to immediately run `weechat-curses`. Using `-c` may prevent loading `.zshrc` (I'm not sure).

mkaito commented:
The first snippet doesn't execute `termite`. It executes `urxvt`, which doesn't have a `-c` flag, and everything after `-e` is just passed to the command to be executed.

From the `urxvt` manpage:

```
-e command [arguments]
           Run the command with its command-line arguments in the urxvt window; also sets the window title
           and icon name to be the basename of the program being executed if neither -title (-T) nor -n are
           given on the command line. If this option is used, it must be the last on the command-line. If
           there is no -e option then the default is to run the program specified by the SHELL environment
           variable or, failing that, sh(1).

           Please note that you must specify a program with arguments. If you want to run shell commands,
           you have to specify the shell, like this:

             urxvt -e sh -c "shell commands"
```

coldfix commented:
> The first snippet doesn't execute termite. It executes urxvt, which doesn't have a -c flag, and everything after -e is just passed to the command to be executed.

My bad, I didn't read properly.

Still, `zsh -c XXX` does not load `.zshrc`, so the behaviour with termite seems as expected. Try it out:

```
mv .zshrc .zshrc.bak
echo "echo hello" > .zshrc
zsh -c "echo world"
```

echos only "world", not "hello world"

zsh man page:

```
... Then, if  the shell  is  interactive,  commands are read from /etc/zshrc and then $ZDOTDIR/.zshrc.
```

i.e. zshrc is read only for *interactive* shells.

mkaito commented:
That.... works. Huh. I mean, you're perfectly correct. I'm just confused because `rxvt -name chat -e zsh -c weechat` has worked fine for a very long time. In fact, still does. I wonder what magic `rxvt` does that `termite` doesn't...

Okay, now that it works, I really want this merged and released :+1:

coldfix commented:
Hi, @jelly 

I could change this PR to only add the new syntax for now, and then we think about whether `-e` should be removed in a second step, if that makes it any easier? In this case, we wouldn't have to worry about backward incompatibility for now and the `-e` option could be deprecated for one or two versions before removing it.

Best, Thomas

frebib commented:
Bump?

mkaito commented:
So... how's it going? :smile:

Eluminae commented:
Guys, it is far later but I'm still having lots of problems using termite on .desktop entries cause of behavior difference with other terminal emulator.

Is there anyone having the same issues ?
I'd like to not fall back to xterm

Eluminae commented:
I'm of course speaking about -e parameters handling

varac commented:
Ping, any updates on this ?

thestinger commented:
Termite is obsolete. Please use [Alacritty](https://github.com/alacritty/alacritty) instead. See https://github.com/thestinger/termite#termite-is-obsoleted-by-alacritty for more details.
